api/inc/register_gateway.h

```diff
@@ -17,104 +17,208 @@
+            .address    = (uint32_t) addr, \
+            .value      = 0, \
+            .mask       = msk, \
+            .operation  = op, \
+            .bxlr       = BXLR  \
+        }; \
+        \
+        /* Pointer to the register gateway we just created. The pointer is
+         * located in a discoverable linker section. */ \
+        __attribute__((section(".keep.uvisor.register_gateway_ptr"))) \
+        static uint32_t const register_gateway_ptr = (uint32_t) &register_gateway; \
+        (void) register_gateway_ptr; \
+        \
+        /* Call the actual gateway. */ \
+        uint32_t result = ((uint32_t (*)(void)) ((uint32_t) &register_gateway | 1))(); \
+        result; \
+    })
 
-/* 3 arguments: masked read */
-#define __UVISOR_REGISTER_GATEWAY_METADATA3(src_box, addr, op, mask) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(op)) "\n" \
-    ".word " UVISOR_TO_STRING(mask) "\n" \
-    "skip_args%=:\n"
+/** Register Gateway - Write operation
+ *
+ * This macro provides an API to perform 32-bit write operations on restricted
+ * registers. Such accesses are assembled into a read-only flash structure that
+ * is read and validated by uVisor before performing the operation.
+ *
+ * @warning These APIs currently only support link-time known values. This means
+ * that an access of this type is not supported:
+ * ```C
+ * static uint32_t const const_value = 1;
+ * uvisor_write(&SYSCFG->CTRL, const_value, UVISOR_RGW_OP_WRITE, 0xF); // Yes
+ * uint32_t var_value = rand();
+ * uvisor_write(&SYSCFG->CTRL, var_value, UVISOR_RGW_OP_WRITE, 0xF); // No
+ * ```
+ *
+ * @param box_name[in]  The name of the source box as decalred in
+ *                      `UVISOR_BOX_CONFIG`.
+ * @param addr[in]      The address for the data access.
+ * @param val[in]       The value to write at address.
+ * @param operation[in] The operation to perform at the address for the read. It
+ *                      is chosen among the `UVISOR_RGW_OP_*` macros.
+ * @param mask[in]      The mask to apply for the write operation.
+ */
+#define uvisor_write(src_box, addr, val, op, msk) \
+    ({ \
+        /* Instanstiate the gateway. This gets resolved at link-time. */ \
+        static TRegisterGateway const register_gateway = { \
+            .svc_opcode = UVISOR_SVC_OPCODE(UVISOR_SVC_ID_REGISTER_GATEWAY), \
+            .branch     = BRANCH_OPCODE(OFFSETOF(TRegisterGateway, branch), OFFSETOF(TRegisterGateway, bxlr)), \
+            .magic      = UVISOR_REGISTER_GATEWAY_MAGIC, \
+            .box_ptr    = (uint32_t) &box_name ## _cfg_ptr, \
+            .address    = (uint32_t) addr, \
+            .value      = val, \
+            .mask       = msk, \
+            .operation  = op, \
+            .bxlr       = BXLR  \
+        }; \
```

Comment:
Note that you didn't need to add `__attribute__((aligned(2))` here, because you are doing `| 1` and the object might (slight chance) not be aligned to a 2 byte boundary in which case no change to the address of the object needs to change for it to be a valid thumb-2 target.